haystack/agents/conversational.py

```diff
@@ -60,3 +61,75 @@ def __init__(
             else lambda query, agent, **kwargs: {"query": query, "history": agent.memory.load()},
             final_answer_pattern=r"^([\s\S]+)$",
         )
+
+
+class ConversationalAgentWithTools(Agent):
+    """
+    A ConversationalAgentWithTools is an extension of the Agent class that supports the use of tools in
+    conversational chat applications. This agent can make use of the ToolsManager to manage a set of tools
+    and seamlessly integrate them into the conversation.
+
+    Example usage:
+
+    ```
+    import os
+
+    from haystack.agents.base import ConversationalAgentWithTools
+    from haystack.agents.memory import ConversationSummaryMemory
+    from haystack.nodes import PromptNode
+    from haystack.tools import ToolsManager, Tool
+
+    # Initialize a PromptNode and a ToolsManager with the desired tools
+    pn = PromptNode("gpt-3.5-turbo", api_key=os.environ.get("OPENAI_API_KEY"), max_length=256)
+    tools_manager = ToolsManager([Tool(name="ExampleTool", pipeline_or_node=example_tool_node)])
+
+    # Create the ConversationalAgentWithTools instance
+    agent = ConversationalAgentWithTools(pn, tools_manager=tools_manager)
+
+    # Use the agent in a chat application
+    while True:
+        user_input = input("Human (type 'exit' or 'quit' to quit): ")
+        if user_input.lower() == "exit" or user_input.lower() == "quit":
+            break
+        else:
+            assistant_response = agent.run(user_input)
+            print("\nAssistant:", assistant_response)
+    ```
+    """
+
+    def __init__(
+        self,
+        prompt_node: PromptNode,
+        prompt_template: Optional[Union[str, PromptTemplate]] = None,
```

Comment:
Heads up that this makes the init of this class different to `ConversationalAgent` with this design: the prompt node and the prompt template are given separately here. Which is easier for initializing the class, and what we have for Agent rn too.

Can we at some point revisit removing prompt node as an init parameter for agent? 🙏

Reply:
Spoke with @julian-risch - what makes most sense when there's time is probably to add `prompt_template` to the init of `ConversationalAgent` too

Reply:
I rebased this PR now, let's wait for green CI and then we can start merging PR one by one